docs/tech/rfc/0019-240310-cesium-race-conditions.md

```diff
@@ -7,129 +7,616 @@
+
+A write proceeds in three phases:
+1. Acquire a writer (getting a file handle to an existing file or opening a new one)
+to an underlying file that still has space for more data.
+2. Write the binary telemetry data to the file handle.
+3. Commit the writer, i.e. add a pointer in the index storing the time range represented
+by the data and the location of the data in the file system (file key, offset, length).
+
+Cesium is different from other Database systems in that a writer may be
+committed multiple times: a writer may repeat steps 2 and 3, i.e.
+
+4. Write more binary telemetry data to the file handle
+5. Commit the writer, i.e. update the pointer that previously described the domain
+written by this writer to contain the new domain.
+
+In Synnax version 19, file cutoffs were introduced, which upon a writer writing to a file
+that reaches its size limit, automatically ends the domain and begins a new one. We must
+also consider the creation of this new pointer when analyzing for race conditions.
+
+#### 4.2.1.1 Acquiring a writer
+
+When acquiring a writer, we must guarantee that we acquire a file handle on a file that
+is both exclusive and under the file size limit. We first scan through the open file handles
+in the file controller under a `RLock` and upon finding a file within the size limit, we
+try to acquire it by comparing-and-swapping its flag (`True` indicates the writer is currently
+in use) from `False` to `True`. If successfully acquired, we return that file handle as
+our writer and the acquisition of a writer is complete.
+Note that although `RLock` does not guarantee exclusivity, the atomic flag
+guarantees that only one writer will ever be in control of a file handle.
+
+If none were acquired out of the open file handles, we first calculate the number of file
+handles in the file controller under an `RLock`: if it is below the limit, then we may
+acquire a new writer – note that there is a possible race condition here (#1): multiple
+callers may each try to create a new writer after checking that the number of file handles
+is below the limit, pushing the total number of file handles beyond the limit.
+
+In the case where we open a new writer, we lock the mutex of the writer completely with
+an exclusive `Lock` and release it when we have a new writer. The acquisition of a writer
+is complete.
+
+In the case where we do not open a new writer, i.e. the number of file descriptors exceeds
+the limit, we garbage-collect all writers (file handles to oversize files) under an exclusive
+lock, and try to acquire a writer again if we garbage-collected any writers. If we did
+not, then we wait for a writer to be released (signaled by an input into the channel `release`),
+and try to acquire a writer again.
+
+#### 4.2.1.2 Writing data with a writer
+
+`Write` writes binary data into the file handle acquired in the previous step and updates
+the `len` and `fileSize` fields of the writer. Note that `Write` may not be called
```

Comment:
Again, there's more code like "updating the `len` and `fileSize` field of the writer". Could you start of this section with a description of the writer and it's fields?